### PR TITLE
Added selection classes

### PR DESCRIPTION
**Pull request by matija-hustic**
Option 3 from #7469 that would support the type of functionality proposed by @MustaphaTR:

http://ares-developers.github.io/Ares-docs/new/typeselect.html

Basically whenever the standard select-by-type functionality is supposed to be extended to select any custom class of selectable actors, these custom classes can be defined using the Class property of the Selectable trait.

If the basic idea is fine, I'd define any remaining selection classes in the YAMLs.

### CODE REVIEW

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 1]
```diff
@@ -244,7 +249,7 @@ public override bool HandleKeyPress(KeyInput e)
 						Game.Debug("Selected across screen");
 					else
 					{
-						newSelection = World.ActorMap.ActorsInWorld().Where(cond);
+						newSelection = World.ActorMap.ActorsInWorld().Where(x => x.HasTrait<Selectable>() && x.Trait<Selectable>().Info.Selectable && cond(x));
 						Game.Debug("Selected across map");
 					}
```
Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: I was thinking about putting the select-by-type logic inside a separate function since these seem similar enough to allow for exporting the functionality

**Comment by Phrohdoh** [comment]
The class ought to default to the actor's type name IMO.

**Comment by matija-hustic** [comment]
Done that @Phrohdoh. Simplified stuff a lot.

**Comment by Phrohdoh** [comment]
That means you can remove the redundant yaml lines.

**Comment by matija-hustic** [comment]
Not sure what you mean... Defaulting selection class to actor name only simplified the code. Hasn't changed anything for what is needed in YAMLs.

Did you mean defaulting on the Tooltip.Name instead of the actor name?

File: OpenRA.Game/Traits/Selectable.cs [hunk 2]
```diff
@@ -22,11 +22,16 @@ public class SelectableInfo : ITraitInfo
 		public readonly int[] Bounds = null;
 		[VoiceReference] public readonly string Voice = null;
 
+		[Desc("All units having the same selection class specified will be selected with select-by-type commands (e.g. double-click).")]
```
Comment (penev92) [inline OpenRA.Game/Traits/Selectable.cs]: You should mention that it defaults to the actor's name.

**Comment by penev92** [comment]
YAML changes for D2k vehicles will be needed. I'm not sure about others (the RA cases you added), though.

**Comment by matija-hustic** [comment]
I'll remove those. So guys, opinions on using Tooltip.Name as default and actor name as fallback in case actor doesn't implement Tooltip or doesn't define Tooltip.Name? That way even those YAML entries for d2k would be rendered unnecessary. Although wouldn't be the cleanest of solutions...

**Comment by Phrohdoh** [comment]
Just stick with the actor type as the default. This isn't exposed to the player so doesn't need to be (and shouldn't be coupled with) something as 'pretty' as the tooltip name.

File: mods/d2k/rules/starport.yaml [hunk 3]
```diff
@@ -21,7 +21,6 @@ trike.starport:
 	Inherits: trike
 	Buildable:
 		Queue: Starport
-		Prerequisites: starport
```
Comment (penev92) [inline mods/d2k/rules/starport.yaml]: This looks out of place in this PR?

**Comment by penev92** [comment]
If I understand the whole idea correctly, the `Selectable: Class:` should go on the startport actors *and maybe* on the normal ones for clarity.

**Comment by matija-hustic** [comment]
Yeah. See here

http://ares-developers.github.io/Ares-docs/new/typeselect.html

File: mods/d2k/rules/starport.yaml (hunk 3 above)
Comment (matija-hustic) [inline mods/d2k/rules/starport.yaml]: Yes. Must make a separate PR?

Comment (penev92) [inline mods/d2k/rules/starport.yaml]: I don't understand why you're removing it to begin with.

Comment (matija-hustic) [inline mods/d2k/rules/starport.yaml]: It is redundant. Compare with the other entries - if they're purchasable in the Starport queue, which is only activated once you actually build the Starport, there is no need to put starport in the prerequisites.

Comment (penev92) [inline mods/d2k/rules/starport.yaml]: Oh, OK, sorry, I see your point.

**Comment by Phrohdoh** [comment]
You can use a Defaults entry or a parent definition if necessary. This is at the yaml level now though.

Above I was referring to engine plumbing.

**Comment by penev92** [comment]
OK, I have no idea why you added those classes in RA, but they seem redundant.
Also D2k needs them on the actor definitions in `starport.yaml`.

**Comment by matija-hustic** [comment]
The `Selectable` trait gets inherited and so does the `Class` property. That way, anyone who inherits from e.g. `trike` will inherit share the same `Selectable.Class` thus making them equivalent as far as selectability is concerned.
I've added this thing for the civilians in order for them all to be in the same class. Try making a selling a couple of Tech Labs and see how fun it is when you get all kinds of different civilians and no quick way to select them all.

**Comment by penev92** [comment]
Ah, good thinking with the inheritance, although that makes it quite implicit.

**Comment by matija-hustic** [comment]
I can add a warning to the description of the `Class` property?

**Comment by penev92** [comment]
No, it's good. I just forgot that the starport actors inherit from those. I guess only thing left is to test ingame now.

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 4]
```diff
@@ -111,10 +111,10 @@ public override bool HandleMouseInput(MouseInput mi)
 
 						if (unit != null && unit.Owner == (World.RenderPlayer ?? World.LocalPlayer))
 						{
-							var newSelection2 = SelectActorsInBox(World, worldRenderer.Viewport.TopLeft, worldRenderer.Viewport.BottomRight,
-								a => a.Owner == unit.Owner && a.Info.Name == unit.Info.Name);
+							var newSelection = SelectActorsInBox(World, worldRenderer.Viewport.TopLeft, worldRenderer.Viewport.BottomRight,
+								a => a.Owner == unit.Owner && a.Trait<Selectable>().Class == unit.Trait<Selectable>().Class);
```
Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: Repeated trait lookups is not the best thing to do, so add a `var class = unit.Trait<Selectable>().Class;` above and check against that here.

**Comment by matija-hustic** [comment]
Fixed commented code @penev92.

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 5]
```diff
@@ -236,20 +236,18 @@ public override bool HandleKeyPress(KeyInput e)
 				}
 				else if (key == Game.Settings.Keys.SelectUnitsByTypeKey)
 				{
-					var selectedTypes = World.Selection.Actors
+					var selectedClasses = World.Selection.Actors
 						.Where(x => x.Owner == player)
-						.Select(a => a.Info);
+						.Select(a => a.Trait<Selectable>().Class);
```
Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: `.ToHashSet()` please

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 6]
```diff
@@ -236,20 +236,18 @@ public override bool HandleKeyPress(KeyInput e)
...
 
-					Func<Actor, bool> cond = a => a.Owner == player && selectedTypes.Contains(a.Info);
-					var tl = worldRenderer.Viewport.TopLeft;
-					var br = worldRenderer.Viewport.BottomRight;
-					var newSelection = SelectActorsInBox(World, tl, br, cond);
+					Func<Actor, bool> cond = a => a.Owner == player && selectedClasses.Contains(a.Trait<Selectable>().Class);
+					var newSelection = SelectActorsInBox(World, worldRenderer.Viewport.TopLeft, worldRenderer.Viewport.BottomRight, cond);
```
Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: `.ToList()`/`.ToArray`

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 7]
```diff
@@ -236,20 +236,18 @@ public override bool HandleKeyPress(KeyInput e)
...
 
-					if (newSelection.Count() > selectedTypes.Count())
+					if (newSelection.Count() > selectedClasses.Count())
```
Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: This makes no sense, and will likely break when you add `.ToHashSet()` above.
I think this should be `newSelection.Count() > World.Selection.Actors.Count()`. Can someone confirm?

Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: `selectedClasses` contains one `string` object per selected unit before invoking select-by-key.
`newSelection` contains one `Actor` object per selected unit after all actors on the screen with the same selection class have been selected.

`selectedClasses.Count()` gives exactly the number of units selected before selecting everyone on the screen
`newSelection.Count()` gives exactly the number of units selected after selecting everyone on the screen

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 8]
```diff
@@ -236,20 +236,18 @@ public override bool HandleKeyPress(KeyInput e)
...
 						Game.Debug("Selected across screen");
 					else
 					{
-						newSelection = World.ActorMap.ActorsInWorld().Where(cond);
+						newSelection = World.ActorMap.ActorsInWorld().Where(x => x.HasTrait<Selectable>() && x.Trait<Selectable>().Info.Selectable && cond(x));
```
Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: `cond` already does the relevant checks, so these additional ones aren't necessary.
It's worth mentioning that by using `.Trait<Selectable>()` everywhere you're walking a fine line, but I'm counting that the owner check will filter out any unselectable actors, and this will only crash when there is bogus YAML (an actor belongs to the player but doesn't have the `Selectable` trait).

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs (hunk 7 above)
Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: `selectedClasses` does in fact contain duplicates, but so has the previous implementation... I've just left it as it is (being conservative).

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs (hunk 8 above)
Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: These checks are necessary since players can own non-selectable units.

**Comment by penev92** [comment]
Adding the "Changes requested" label as discussed above and on IRC.

**Comment by matija-hustic** [comment]
Fixed as @penev92 requested.

**Comment by matija-hustic** [comment]
@penev92, after we talked in IRC, I felt more comfortable changing other things that seem that could use some housekeeping. Could you check if this is more to your liking?

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 9]
```diff
@@ -264,18 +270,41 @@ public override bool HandleKeyPress(KeyInput e)
...
+			return sel.Where(a =>
+			{
+				if (a.Owner != player)
+					return false;
+				var s = a.TraitOrDefault<Selectable>();
+
+				// sc == null means that units, that meet all other criteria, get selected
+				return s != null && s.Info.Selectable && (sc == null || sc.Contains(s.Class));
+			});
+		}
```
Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: I can probably change something with the above three functions, but too tired right now to think clearly.

Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: Can `sc` really be `null` here?

Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: Think caryalls in d2k.

**Comment by penev92** [comment]
Code looks good, and it works ingame :+1: 
You should squash the commits now.

**Comment by matija-hustic** [comment]
Squashed.

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs (hunk 9 above)
Comment (Phrohdoh) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: Can we get better variable names please?

**Comment by Mailaender** [comment]
Needs a rebase.

**Comment by matija-hustic** [comment]
Changed variable names as @Phrohdoh suggested. Rebased.

File: OpenRA.Game/Traits/Selectable.cs [hunk 10]
```diff
@@ -33,6 +38,10 @@ public Selectable(Actor self, SelectableInfo info)
 		{
 			this.self = self;
 			Info = info;
+			if (string.IsNullOrEmpty(info.Class))
```
Comment (Phrohdoh) [inline OpenRA.Game/Traits/Selectable.cs]: This could become ternary, but that doesn't change the actual logic so it isn't necessary.

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 11]
```diff
@@ -111,15 +113,17 @@ public override bool HandleMouseInput(MouseInput mi)
 
 						if (unit != null && unit.Owner == (World.RenderPlayer ?? World.LocalPlayer))
 						{
-							var newSelection2 = SelectActorsInBox(World, worldRenderer.Viewport.TopLeft, worldRenderer.Viewport.BottomRight,
-								a => a.Owner == unit.Owner && a.Info.Name == unit.Info.Name);
+							// Select actors on the screen that have the same selection class as the already selected actor
+							var sc = unit.Trait<Selectable>().Class;
```
Comment (Phrohdoh) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: Are you ensuring that `unit` has `Selectable` anywhere?

File: mods/ra/rules/infantry.yaml [hunk 12]
```diff
@@ -388,6 +388,8 @@ DELPHI:
 
 CHAN:
 	Inherits: ^CivInfantry
+	Selectable:
+		Class: CHAN
```
Comment (Phrohdoh) [inline mods/ra/rules/infantry.yaml]: From my understanding you don't need to define `Class` if you want to use the actor definition's name, which is what this is doing.

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs (hunk 11 above)
Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: Nope. Nice bug spotting there.

Comment (Phrohdoh) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: There is a similar case below.

Comment (penev92) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: See if ScreenMap.WithHighestSelectionPriority() doesn't guarantee that
already.

File: mods/ra/rules/infantry.yaml (hunk 12 above)
Comment (matija-hustic) [inline mods/ra/rules/infantry.yaml]: The point here is that inheritance overrides default `Class` attribution.
Here `^CivInfantry` already defines the selection class for `CHAN`, `GNRL` and other generic civilian actors, which they inherit. But `CHAN`, `GNRL` should not use this inherited `Class`, but should be different. See the example of the combat tank in D2K - the actual `Class` property is defined for only `^combat` but gets inherited by six different actors (3 for each race times 2 for normal / starport).

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs (hunk 11 above)
Comment (matija-hustic) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: It doesn't @penev92, I've checked. I have to do the check.
@Phrohdoh, I don't see a similar case below.

File: mods/ra/rules/infantry.yaml (hunk 12 above)
Comment (Phrohdoh) [inline mods/ra/rules/infantry.yaml]: I hadn't noticed that, nice. :100:

File: OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs [hunk 13]
```diff
@@ -230,26 +237,28 @@ public override bool HandleKeyPress(KeyInput e)
...
 					World.Selection.Combine(World, ownUnitsOnScreen, false, false);
 				}
 				else if (key == Game.Settings.Keys.SelectUnitsByTypeKey)
 				{
-					var selectedTypes = World.Selection.Actors
+					// Get all the selected actors' selection classes
+					var selectedClasses = World.Selection.Actors
 						.Where(x => x.Owner == player)
-						.Select(a => a.Info);
+						.Select(a => a.Trait<Selectable>().Class)
```
Comment (Phrohdoh) [inline OpenRA.Game/Widgets/WorldInteractionControllerWidget.cs]: This.

**Comment by Phrohdoh** [comment]
Please squash the `Typos` commit so this can be tested.

**Comment by matija-hustic** [comment]
Squashed.

**Comment by abcdefg30** [comment]
Looks also good to me. :+1: / :white_check_mark:

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/cb45ff5db15d738468c3e3c2320d35b1cbba5121...c043334d6197d7a3d4cbb51333b66f0aea0458e0)

**Comment by penev92** [comment]
[Changelog #2](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/c043334d6197d7a3d4cbb51333b66f0aea0458e0...71b56bb0105876a84bad1e87459f017dbeed4cdc)
